core/lib/types/src/storage/writes/mod.rs

```diff
@@ -188,6 +188,78 @@ fn prepend_header(compressed_state_diffs: Vec<u8>) -> Vec<u8> {
     res.to_vec()
 }
 
+/// Struct for storing tree writes in DB.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub struct TreeWrite {
+    /// `address` part of storage key.
+    pub address: Address,
+    /// `key` part of storage key.
+    pub key: H256,
+    /// Value written.
+    pub value: H256,
+    /// Leaf index of the slot.
+    pub leaf_index: u64,
+}
+
+impl TreeWrite {
+    const TREE_WRITE_BYTES_SIZE: usize = 20 + 32 + 32 + 8;
+
+    fn extend_with_bytes(&self, vec: &mut Vec<u8>) {
+        vec.extend_from_slice(&self.address.0);
+        vec.extend_from_slice(&self.key.0);
+        vec.extend_from_slice(&self.value.0);
+        vec.extend_from_slice(&self.leaf_index.to_be_bytes());
+    }
+
+    pub fn from_slice(slice: &[u8]) -> anyhow::Result<Self> {
+        if slice.len() == Self::TREE_WRITE_BYTES_SIZE {
+            Ok(Self {
+                address: Address::from_slice(&slice[0..20]),
+                key: H256::from_slice(&slice[20..52]),
+                value: H256::from_slice(&slice[52..84]),
+                leaf_index: u64::from_be_bytes(slice[84..92].try_into().unwrap()),
+            })
+        } else {
+            anyhow::bail!(
+                "Incorrect slice len, expected: {}, got {}",
+                Self::TREE_WRITE_BYTES_SIZE,
+                slice.len()
+            );
+        }
+    }
+
+    pub fn vec_to_bytes(vec: Vec<Self>) -> Vec<u8> {
+        let size = Self::TREE_WRITE_BYTES_SIZE * vec.len();
+        let mut result = Vec::with_capacity(size);
+
+        for element in vec {
+            element.extend_with_bytes(&mut result);
+        }
+
+        result
+    }
```

Comment:
I saw that you switched the implementation to use manual (de)serialization but the cause is still unclear to me. I assumed that `bincode` would product efficient implementations for `fixed-hash` types, is that not the case in your experience? Also, even if you can produce more efficient implementation, is there a reason not to implement `Serialize`/`Deserialize` traits instead of ad-hoc functions (there would be a miniscule overhead for vec length but otherwise I expect the behaviour to be identical)?

Reply:
Yes, default serialization for fixed-hash types always serializes to strings, meaning 2x bytes then needed, see https://github.com/paritytech/parity-common/blob/master/primitive-types/impls/serde/src/serialize.rs.

As for `Serialize/Deserialize` impls, makes sense, I will try that